bpf/lib/host_firewall.h

```diff
@@ -331,24 +417,50 @@ ipv4_host_policy_ingress(struct __ctx_buff *ctx, __u32 *src_id,
 
 	/* Only enforce host policies for packets to host IPs. */
 	if (dst_id != HOST_ID)
-		return CTX_ACT_OK;
+		return false;
 
 	/* Lookup connection in conntrack map. */
-	tuple.nexthdr = ip4->protocol;
-	tuple.daddr = ip4->daddr;
-	tuple.saddr = ip4->saddr;
+	tuple->nexthdr = ip4->protocol;
+	tuple->daddr = ip4->daddr;
+	tuple->saddr = ip4->saddr;
 	l4_off = l3_off + ipv4_hdrlen(ip4);
-#  ifndef ENABLE_IPV4_FRAGMENTS
-	/* Indicate that this is a datagram fragment for which we cannot
-	 * retrieve L4 ports. Do not set flag if we support fragmentation.
-	 */
-	is_untracked_fragment = ipv4_is_fragment(ip4);
-#  endif
-	ret = ct_lookup4(get_ct_map4(&tuple), &tuple, ctx, l4_off, CT_INGRESS,
-			 &ct_state, &trace->monitor);
-	if (ret < 0)
-		return ret;
+	ct_buffer->ret = ct_lookup4(get_ct_map4(tuple), tuple, ctx, l4_off, CT_INGRESS,
+				    &ct_buffer->ct_state, &ct_buffer->monitor);
 
+	return true;
+}
+
+static __always_inline int
+__ipv4_host_policy_ingress(struct __ctx_buff *ctx, struct iphdr *ip4,
+			   struct ct_buffer4 *ct_buffer, __u32 *src_id,
+			   struct trace_ctx *trace)
+{
+	struct ct_state ct_state_new = {};
+#ifndef HAVE_DIRECT_ACCESS_TO_MAP_VALUES
+	struct ct_state ct_state_on_stack;
+	struct ipv4_ct_tuple tuple_on_stack;
+#endif
+	struct ct_state *ct_state;
+	struct ipv4_ct_tuple *tuple;
+	int ret = ct_buffer->ret;
+	int verdict = CTX_ACT_OK;
+	__u8 policy_match_type = POLICY_MATCH_NONE;
+	__u8 audited = 0;
+	struct remote_endpoint_info *info;
+	bool is_untracked_fragment = false;
+	__u16 proxy_port = 0;
+
+#ifdef HAVE_DIRECT_ACCESS_TO_MAP_VALUES
+	ct_state = &ct_buffer->ct_state;
+	tuple = &ct_buffer->tuple;
+#else
+	memcpy(&tuple_on_stack, &ct_buffer->tuple, sizeof(tuple_on_stack));
+	tuple = &tuple_on_stack;
+	memcpy(&ct_state_on_stack, &ct_buffer->ct_state, sizeof(ct_state_on_stack));
+	ct_state = &ct_state_on_stack;
+#endif
```

Comment:
Note that this will "conflict" with https://github.com/cilium/cilium/pull/24504.